### ISSUE TITLE
Is there a way to set additional attributes like "application/json" on the clipboard object?

### ISSUE THREAD

**Issue by tnrich** [zenorocha/clipboard.js#484]
Hey there,

Thanks for this awesome library! 

I was wondering if it might be possible to set additional metadata to be copied. Usually that is done using `event.clipboardData.setData('application/json', "{"hey":"there"}")` 

https://w3c.github.io/clipboard-apis/#override-copy

Is it possible to do that using this library?

**Comment by zenorocha** [comment]
Hey @tnrich,

We don't have that option, but you could totally use something like `JSON.stringify`.

**Comment by DelfsEngineering** [comment]
@tnrich Were you ever able to do this? I want to place application/xml on the clipboard. Are there security blocks in place?

@zenorocha  Serializing the JSON is not the same as setting the format of the contents.

**Comment by tnrich** [comment]
You can do it by listening for the copy event manually. This is kind of funky but works: 

```
new Clipboard(`.${className}`, {
          text: () => {
            this.sequenceDataToCopy = sequenceDataToCopy;
            document.body.addEventListener("copy", this.handleCopy);
            return sequenceDataToCopy.sequence;
          }
        });
```

In handleCopy just manually add the special metadata:

```
const clipboardData = event.clipboardData;
          clipboardData.setData("text/plain", copiedSequenceData.sequence);
          clipboardData.setData(
            "application/json",
            JSON.stringify(copiedSequenceData)
          );
```

Sorry about the random irrelevant variable names I am on my phone.. :)

Lemme know if that works for you

**Comment by crymis** [comment]
I have the same issue. I need to copy a document as `text/xml`.
I am developing for IE11, which makes it a tough one...

The `event.clipboardData.setData` method is apparently undefined when accessing. I already found the `setTimeout`-hack, but it does not work for me. 

Any suggestions? Thanks in advance!

**Comment by tnrich** [comment]
@crymis I am not familiar because we're not developing for IE. Google away my friend ...

**Comment by crymis** [comment]
@tnrich I forgot to tell you, that I come to this issue while googling for the IE problem and tried your solution. Unfortunately it didn't work. 
Maybe I don't understand the variable names correctly but in my attempt with React it looked like that (tested in Chrome):

```	
private clipboardOrderData: ClipboardJS;
private sequenceDataToCopy: string = 'empty!';

componentDidMount() {
	this.clipboardData = new ClipboardJS('.copyData', {
		text: () => {
			this.sequenceDataToCopy = new DOMParser().parseFromString('<?xml version="1.0" ?><root />', 'text/xml');
			document.body.addEventListener('copy', this.handleCopy);
			return 'not my xml';
		},
	});
}

handleCopy = (event: ClipboardEvent) => {
	const clipboardData = event.clipboardData;
	clipboardData.setData('text/plain', this.sequenceDataToCopy.nodeName);
	clipboardData.setData('text/xml', new XMLSerializer().serializeToString(this.sequenceDataToCopy));
}

componentWillUnmount() {
	this.clipboardData.destroy();
}
```

And this Button in the render-function:
`<Button className="copyData">Copy me</Button>`

When i paste the result after clicking the button, 'not my xml' is printed.

Could you explain the process of the `text` function?

**Comment by tnrich** [comment]
Can you listen to the onPaste event and see if you're setting the text/xml correctly? My understanding of the text function is that whatever is returned from it will be copied to the user's clipboard. Looking at it now it would seem that setting the text/plain in the handleCopy event should override that though. On my phone so I'll need to do some more investigating and get back to you on why you would still be getting "not my xml" copied..

**Comment by crymis** [comment]
I added a listener for the paste event. This is what happens:

![clipboardpastelistener](https://user-images.githubusercontent.com/3726683/38600928-2f10b294-3d66-11e8-972d-3df41088564a.png)

So, when copying there is no type set in the event and it's `text/plain` in the paste event. It seems there is nothing overwritten. 

I think adding the copy-listener in the text() function is too late. The return command is immediately called.

The click on the button fires the copy event which triggers the text() function. Adding the listener inside that text() function could may change the following copy event, but not the current.

I already played around and tried some things, without success.
Tell me, if you find out what's wrong here.
